cmd/operator/main.go

```diff
@@ -101,13 +102,14 @@ func main() {
 		pm                    *patcher.Manager
 	)
 
-	ctrl.SetLogger(klogr.New())
+	ctrl.SetLogger(klogr.NewWithOptions(klogr.WithFormat(klogr.FormatKlog)))
 
 	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
 	flag.StringVar(&devicePluginNamespace, "deviceplugin-namespace", metav1.NamespaceSystem, "The namespace where deviceplugin daemonsets are created")
 	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.BoolVar(&controllers.UpgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion))
```

Comment:
I'm still thinking that this is not very good to set module variable this way. Can we add Upgradable call to the controller interface (see example below)? This is more wordy, but gives more flexibility in my opinion. For example, with this approach it's possible to configure only one or subset of plugins to be upgradable. Not sure it makes practical sense though, but it at least makes sense to discuss it. @mythi @rojkov @ozhuraki  WDYT?
```
diff --git a/cmd/operator/main.go b/cmd/operator/main.go
index c25637c6..66f25be6 100644
--- a/cmd/operator/main.go
+++ b/cmd/operator/main.go
@@ -57,7 +57,7 @@ func init() {
        // +kubebuilder:scaffold:scheme
 }
 
-type devicePluginControllerAndWebhook map[string](func(ctrl.Manager, string, bool) error)
+type devicePluginControllerAndWebhook map[string](func(ctrl.Manager, string, bool, bool) error)
 
 type flagList []string
 
@@ -100,6 +100,7 @@ func main() {
                devicePluginNamespace string
                enableLeaderElection  bool
                pm                    *patcher.Manager
+               upgradePlugins        bool
        )
 
        ctrl.SetLogger(klogr.NewWithOptions(klogr.WithFormat(klogr.FormatKlog)))
@@ -109,7 +110,7 @@ func main() {
        flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
                "Enable leader election for controller manager. "+
                        "Enabling this will ensure there is only one active controller manager.")
-       flag.BoolVar(&controllers.UpgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion))
+       flag.BoolVar(&upgradePlugins, "upgrade", true, fmt.Sprintf("Auto-upgrade plugins' images to the current release %s", controllers.ImageMinVersion))
        flag.Var(&devices, "devices", "Device(s) to set up.")
        flag.Parse()
 
@@ -147,7 +148,7 @@ func main() {
        withWebhook := true
 
        for _, device := range devices {
-               if err = setupControllerAndWebhook[device](mgr, ns, withWebhook); err != nil {
+               if err = setupControllerAndWebhook[device](mgr, ns, withWebhook, upgradePlugins); err != nil {
                        setupLog.Error(err, "unable to initialize controller", "controller", device)
                        os.Exit(1)
                }
diff --git a/pkg/controllers/dlb/controller.go b/pkg/controllers/dlb/controller.go
index e027a710..3244a27c 100644
--- a/pkg/controllers/dlb/controller.go
+++ b/pkg/controllers/dlb/controller.go
@@ -42,8 +42,8 @@ var defaultNodeSelector map[string]string = deployments.DLBPluginDaemonSet().Spe
 // +kubebuilder:rbac:groups=deviceplugin.intel.com,resources=dlbdeviceplugins/finalizers,verbs=update
 
 // SetupReconciler creates a new reconciler for DlbDevicePlugin objects.
-func SetupReconciler(mgr ctrl.Manager, namespace string, withWebhook bool) error {
-       c := &controller{scheme: mgr.GetScheme(), ns: namespace}
+func SetupReconciler(mgr ctrl.Manager, namespace string, withWebhook, upgradable bool) error {
+       c := &controller{scheme: mgr.GetScheme(), ns: namespace, upgradable: upgradable}
        if err := controllers.SetupWithManager(mgr, c, devicepluginv1.GroupVersion.String(), "DlbDevicePlugin", ownerKey); err != nil {
                return err
        }
@@ -57,8 +57,9 @@ func SetupReconciler(mgr ctrl.Manager, namespace string, withWebhook bool) error
 
 type controller struct {
        controllers.DefaultServiceAccountFactory
-       scheme *runtime.Scheme
-       ns     string
+       scheme     *runtime.Scheme
+       ns         string
+       upgradable bool
 }
 
 func (c *controller) CreateEmptyObject() client.Object {
@@ -153,6 +154,10 @@ func (c *controller) UpdateStatus(rawObj client.Object, ds *apps.DaemonSet, node
        return updated, nil
 }
 
+func (c *controller) Upgradable() bool {
+       return c.upgradable
+}
+
 func getPodArgs(gdp *devicepluginv1.DlbDevicePlugin) []string {
        args := make([]string, 0, 4)
        args = append(args, "-v", strconv.Itoa(gdp.Spec.LogLevel))

diff --git a/pkg/controllers/reconciler.go b/pkg/controllers/reconciler.go
index f0e5863c..28e25d81 100644
--- a/pkg/controllers/reconciler.go
+++ b/pkg/controllers/reconciler.go
@@ -36,10 +36,7 @@ import (
        "sigs.k8s.io/controller-runtime/pkg/log"
 )
 
-var (
-       UpgradePlugins bool
-       bKeeper        = &bookKeeper{}
-)
+var bKeeper = &bookKeeper{}
 
 const ImageMinVersion = "0.23.0"
 
@@ -106,6 +103,7 @@ type DevicePluginController interface {
        UpdateDaemonSet(client.Object, *apps.DaemonSet) (updated bool)
        UpdateStatus(client.Object, *apps.DaemonSet, []string) (updated bool, err error)
        Upgrade(ctx context.Context, obj client.Object) (upgrade bool)
+       Upgradable() bool
 }
 
 type reconciler struct {
@@ -226,7 +224,7 @@ func UpgradeImages(image *string, initimage *string) (upgrade bool) {
 
 func upgradeOnce(ctx context.Context, r *reconciler, devicePlugin client.Object) {
        r.once.Do(func() {
-               if UpgradePlugins && r.controller.Upgrade(ctx, devicePlugin) {
+               if r.controller.Upgradable() && r.controller.Upgrade(ctx, devicePlugin) {
                        if err := r.Update(ctx, devicePlugin); err != nil {
                                log := log.FromContext(ctx)
                                log.Error(err, "unable to update devicePlugin")

```

Reply:
I second it. At least it may make testing more convenient and less prone to races. 

Was going to comment on that, but got distracted.

Reply:
@rojkov @mythi @bart0sh 
It boils down to whether the upgrade should be per device or global. At the moment it's fairly simple - global and minimally controllable (on/off, if there's a need).

More fine-grained control:
- Needed in practice?
- Would be simple in understanding and usage?

Not sure. Thoughts?

Reply:
Being global and RW-accessible from anywhere is the problem. `controller.UpgradePlugins` adds global state and introduces a public interface to it which is supposed to be supported *forever*. And we just don't know currently what kind of problems we are going to have down the road. Consider multiple test cases running in parallel in one test process and updating/reading the state. Obviously the cases wouldn't be isolated enough from each other.